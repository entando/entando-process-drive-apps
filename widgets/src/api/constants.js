export const DOMAINS = {
  PDA: process.env.REACT_APP_DOMAIN || '',
  APP_BUILDER: process.env.REACT_APP_APP_BUILDER_DOMAIN || '',
};

export const IS_MOCKED_API = process.env.REACT_APP_MOCKED_API === 'true';

export const MOCK_API_DELAY = 800;

export const LOCAL = process.env.REACT_APP_LOCAL === 'true';

export const METHODS = {
  GET: 'GET',
  POST: 'POST',
  PUT: 'PUT',
  DELETE: 'DELETE',
};

// Inject ur access_token if you are trying to reach the APP-BUILDER API locally
export const TOKENS = {
  APP_BUILDER: '0694821a5bdc2229aac820e8a773669a',
};
