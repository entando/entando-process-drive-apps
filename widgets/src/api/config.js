import { DOMAIN, IS_MOCKED_API, MOCK_API_DELAY } from 'api/constants';
import utils from 'utils';
import config from 'mocks/config';

const get = async widgetCode => {
  if (IS_MOCKED_API) {
    await utils.timeout(MOCK_API_DELAY);
    return config;
  }

  const url = `${DOMAIN}/configs/${widgetCode}`;
  const response = await fetch(url);

  return response.json();
};

export default {
  get,
};
