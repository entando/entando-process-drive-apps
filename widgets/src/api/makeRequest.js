import { IS_MOCKED_API, MOCK_API_DELAY, DOMAINS, METHODS } from 'api/constants';
import utils from 'utils';

const getParams = queryParams =>
  Object.keys(queryParams)
    .filter(k => queryParams[k] !== undefined && queryParams[k] !== '')
    .map(k => `${encodeURIComponent(k)}=${encodeURIComponent(queryParams[k])}`)
    .join('&');

export default async ({
  domain,
  uri,
  method,
  mockResponse = {},
  useAuthentication,
  body,
  headers = {},
  queryParams,
  forceMock,
}) => {
  if (IS_MOCKED_API || forceMock) {
    await utils.timeout(MOCK_API_DELAY);
    return mockResponse;
  }

  const url = `${domain || DOMAINS.PDA}${uri}${queryParams ? `?${getParams(queryParams)}` : ''}`;

  const requestHeaders = useAuthentication
    ? new Headers({
        'Content-Type': 'application/json',
        Authorization: `Bearer ${localStorage.getItem('token')}`,
        ...headers,
      })
    : new Headers(headers);

  const configs = {
    method: method || METHODS.GET,
    ...(body ? { body } : {}),
    headers: requestHeaders,
  };

  const response = await fetch(url, configs);
  if (!response.ok) {
    const payload = await response.json();
    throw new Error(
      payload.message
        ? payload.message
        : `Bad Response from server: ${response.status} ${response.statusText}`
    );
  }

  const responseHeaders = response.headers.get('Content-Type');

  if (responseHeaders && responseHeaders.includes('xml')) {
    return response.text();
  }
  if (responseHeaders && responseHeaders.includes('octet')) {
    return response;
  }
  return response.json();
};
