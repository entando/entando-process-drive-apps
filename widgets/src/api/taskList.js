import { IS_MOCKED_API, MOCK_API_DELAY, SERVICE } from 'api/constants';
import getTasks from 'mocks/taskList/api.mock';
import utils from 'utils';

const get = async (connection, page, pageSize, sortedColumn, sortOrder, filter) => {
  if (IS_MOCKED_API) {
    await utils.timeout(MOCK_API_DELAY);
    return getTasks(page, pageSize, sortedColumn, sortOrder, filter);
  }

  const url = `${SERVICE.URL}/connections/${connection}/tasks?sort=taskId`;
  const token = localStorage.getItem('token');
  const response = await fetch(url, {
    method: 'get',
    headers: new Headers({
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/json',
    }),
  });

  return response.json();
};

export default {
  get,
};
