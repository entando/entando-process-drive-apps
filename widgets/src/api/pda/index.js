import makeRequest from 'api/makeRequest';
import { SERVICE, METHODS } from 'api/constants';

import MOCK_CONNECTIONS from 'mocks/pda/connections';
import MOCK_PROCESSES from 'mocks/pda/processes';

export const getConnections = async () =>
  makeRequest({
    domain: SERVICE.URL,
    uri: '/connections',
    method: METHODS.GET,
    mockResponse: MOCK_CONNECTIONS,
    useAuthentication: false,
  });

export const getProcess = async connection =>
  makeRequest({
    domain: SERVICE.URL,
    uri: `/connections/${connection}/processes/definitions`,
    method: METHODS.GET,
    mockResponse: MOCK_PROCESSES,
    useAuthentication: false,
  });
