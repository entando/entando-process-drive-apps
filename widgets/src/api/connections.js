import { METHODS, DOMAINS } from 'api/constants';

import { CONNECTIONS } from 'mocks/taskList/configs';
import makeRequest from 'api/makeRequest';

// eslint-disable-next-line import/prefer-default-export
export const getConnections = async () =>
  makeRequest({
    domain: DOMAINS.PDA,
    uri: '/connections',
    method: METHODS.GET,
    mockResponse: CONNECTIONS,
    useAuthentication: false,
  });
