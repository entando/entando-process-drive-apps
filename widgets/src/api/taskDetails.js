import { SERVICE, METHODS } from 'api/constants';
import mockedTask from 'mocks/taskDetails/task';

import { CONNECTIONS, PROCESS, GROUPS, COLUMNS } from 'mocks/taskList/configs';
import makeRequest from 'api/makeRequest';

export const getTask = async (connection, taskId) => {
  return makeRequest({
    domain: SERVICE.URL,
    uri: `/connections/${connection}/tasks/${taskId}`,
    method: METHODS.GET,
    mockResponse: mockedTask,
    useAuthentication: true,
  });
};

// Configs

export const getConnections = async () =>
  makeRequest({
    domain: SERVICE.URL,
    uri: '/connections',
    method: METHODS.GET,
    mockResponse: CONNECTIONS,
    useAuthentication: false,
  });

export const getProcess = async connection =>
  makeRequest({
    // domain: PDA_DOMAIN,
    uri: `/connections/${connection}/processes/definitions`,
    method: METHODS.GET,
    mockResponse: PROCESS,
    useAuthentication: true,
  });

export const getGroups = async connection =>
  makeRequest({
    // domain: PDA_DOMAIN,
    uri: `/connections/${connection}/groups`,
    method: METHODS.GET,
    mockResponse: GROUPS,
    useAuthentication: true,
  });

export const getColumns = async connection =>
  makeRequest({
    // domain: PDA_DOMAIN,
    uri: `/connections/${connection}/tasks/columns`,
    method: METHODS.GET,
    mockResponse: COLUMNS,
    useAuthentication: true,
  });
