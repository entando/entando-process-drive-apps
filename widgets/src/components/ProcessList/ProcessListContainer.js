import { ThemeProvider } from '@material-ui/core/styles';
import React from 'react';
import Paper from '@material-ui/core/Paper';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import PropTypes from 'prop-types';
import withStyles from '@material-ui/core/styles/withStyles';

import { DOMAINS, LOCAL } from 'api/constants';
import { getPageWidget } from 'api/app-builder/pages';
import { getProcesses } from 'api/pda/processes';
import theme from 'theme';
import Notification from 'components/common/Notification';
import withAuth from 'components/common/auth/withAuth';
import Table from 'components/common/Table/Table';

const DisplayArray = ({ row }) => (
  <span>{Array.isArray(row.userTasks) ? row.userTasks.join(', ') : row.userTasks}</span>
);

DisplayArray.propTypes = {
  row: PropTypes.shape({
    userTasks: PropTypes.oneOfType([PropTypes.string, PropTypes.array]),
  }).isRequired,
};

const columns = [
  {
    accessor: 'processName',
    header: 'Process Name',
  },
  {
    accessor: 'initiator',
    header: 'Created by',
  },
  {
    accessor: 'date',
    header: 'Created At',
  },
  {
    accessor: 'userTasks',
    header: 'Status',
    customCell: DisplayArray,
  },
];

const styles = {
  toolbar: {
    justifyContent: 'space-between',
    padding: 16,
    minHeight: 'unset',
  },
  title: {
    textAlign: 'left',
  },
};

export class ProcessListContainer extends React.Component {
  state = {
    processes: [],
    loading: false,
    errorAlert: null,
  };

  componentDidMount = async () => {
    const { serviceUrl, pageCode, frameId } = this.props;

    if (!LOCAL) {
      // set the PDA domain to the URL passed via props
      DOMAINS.PDA = serviceUrl;
    }

    try {
      // config will be fetched from app-builder
      const widgetConfigs = await getPageWidget(pageCode, frameId);

      const { config } = widgetConfigs.payload;
      const connection = config.knowledgeSource;

      this.setState({ connection }, this.fetchProcesses);
    } catch (error) {
      this.handleError(error.message, 'messages.errors.dataLoading');
    }
  };

  fetchProcesses = async () => {
    const { connection } = this.state;

    try {
      const processes = await getProcesses(connection);
      this.setState({ processes: processes.payload });
    } catch (error) {
      this.handleError(error.message, 'messages.errors.dataLoading');
    }
  };

  handleError(err) {
    this.setState({
      errorAlert: err.toString(),
      loading: false,
    });
  }

  render() {
    const { loading, processes, errorAlert } = this.state;
    const { classes, title } = this.props;

    return (
      <ThemeProvider theme={theme}>
        <Paper>
          {title && (
            <Toolbar className={classes.toolbar}>
              <div className={classes.title}>
                <Typography variant="h2">Process List</Typography>
              </div>
            </Toolbar>
          )}

          <Table loading={loading} columns={columns} rows={processes} hidePagination />
          <Notification type="error" message={errorAlert} onClose={this.closeNotification} />
        </Paper>
      </ThemeProvider>
    );
  }
}

ProcessListContainer.propTypes = {
  classes: PropTypes.shape({
    toolbar: PropTypes.string,
    title: PropTypes.string,
  }),
  serviceUrl: PropTypes.string,
  pageCode: PropTypes.string,
  frameId: PropTypes.string,
  title: PropTypes.string,
};

ProcessListContainer.defaultProps = {
  classes: {},
  serviceUrl: '',
  pageCode: '',
  frameId: '',
  title: '',
};

export default withAuth(withStyles(styles)(ProcessListContainer), ['process-list']);
