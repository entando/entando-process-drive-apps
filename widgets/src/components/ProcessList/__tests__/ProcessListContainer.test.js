import React from 'react';
import { render, wait } from '@testing-library/react';

import { DOMAINS } from 'api/constants';
import WIDGET_CONFIGS from 'mocks/app-builder/pages';
import MOCKED_RESPONSE from 'mocks/pda/processList';
import { ProcessListContainer } from '../ProcessListContainer';
import 'mocks/i18nMock';

describe('<ProcessListContainer />', () => {
  it('renders snapshot correctly', async () => {
    const configUrl = `${DOMAINS.APP_BUILDER}/api/pages//widgets/`;
    const connection = 'kieStaging';
    const processList = `/connections/${connection}/processes/instances?`;

    fetch.once(JSON.stringify(WIDGET_CONFIGS.PROCESS_LIST)).once(JSON.stringify(MOCKED_RESPONSE));

    const { container } = render(<ProcessListContainer />);

    await wait(() => expect(container).toMatchSnapshot());

    expect(fetch.mock.calls.length).toBe(2);
    expect(fetch.mock.calls[0][0]).toEqual(configUrl);
    expect(fetch.mock.calls[1][0]).toEqual(processList);
  });
});
