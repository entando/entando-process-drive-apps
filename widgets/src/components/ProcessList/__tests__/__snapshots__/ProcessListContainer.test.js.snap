// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProcessListContainer /> renders snapshot correctly 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    
    <div
      class="Table-tableWrapper-30"
    >
      <table
        class="MuiTable-root Table-hideShadows-29"
      >
        <thead
          class="MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head StyledHeaderCell-head-39"
              scope="col"
              style="white-space: nowrap;"
            >
              Process Name
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head StyledHeaderCell-head-39"
              scope="col"
              style="white-space: nowrap;"
            >
              Created by
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head StyledHeaderCell-head-39"
              scope="col"
              style="white-space: nowrap;"
            >
              Created At
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head StyledHeaderCell-head-39"
              scope="col"
              style="white-space: nowrap;"
            >
              Status
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root StyledTableRowHover-root-91 MuiTableRow-hover"
            style="height: 55px; cursor: initial;"
          >
            <td
              class="MuiTableCell-root StyledTableCell-root-93 MuiTableCell-body MuiTableCell-sizeSmall"
            >
              MortgageApprovalProcess
            </td>
            <td
              class="MuiTableCell-root StyledTableCell-root-93 MuiTableCell-body MuiTableCell-sizeSmall"
            >
              pamAdmin
            </td>
            <td
              class="MuiTableCell-root StyledTableCell-root-93 MuiTableCell-body MuiTableCell-sizeSmall"
            >
              2020-04-09T11:53:52.999
            </td>
            <td
              class="MuiTableCell-root StyledTableCell-root-93 MuiTableCell-body MuiTableCell-sizeSmall"
            >
              <span>
                Qualify, test
              </span>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root StyledTableRowHover-root-91 MuiTableRow-hover"
            style="height: 55px; cursor: initial;"
          >
            <td
              class="MuiTableCell-root StyledTableCell-root-93 MuiTableCell-body MuiTableCell-sizeSmall"
            >
              MortgageApprovalProcess
            </td>
            <td
              class="MuiTableCell-root StyledTableCell-root-93 MuiTableCell-body MuiTableCell-sizeSmall"
            >
              pamAdmin
            </td>
            <td
              class="MuiTableCell-root StyledTableCell-root-93 MuiTableCell-body MuiTableCell-sizeSmall"
            >
              2020-04-09T11:53:55.557
            </td>
            <td
              class="MuiTableCell-root StyledTableCell-root-93 MuiTableCell-body MuiTableCell-sizeSmall"
            >
              <span>
                Qualify
              </span>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root StyledTableRowHover-root-91 MuiTableRow-hover"
            style="height: 55px; cursor: initial;"
          >
            <td
              class="MuiTableCell-root StyledTableCell-root-93 MuiTableCell-body MuiTableCell-sizeSmall"
            >
              MortgageApprovalProcess
            </td>
            <td
              class="MuiTableCell-root StyledTableCell-root-93 MuiTableCell-body MuiTableCell-sizeSmall"
            >
              pamAdmin
            </td>
            <td
              class="MuiTableCell-root StyledTableCell-root-93 MuiTableCell-body MuiTableCell-sizeSmall"
            >
              2020-05-07T09:38:13.505
            </td>
            <td
              class="MuiTableCell-root StyledTableCell-root-93 MuiTableCell-body MuiTableCell-sizeSmall"
            >
              <span>
                Qualify
              </span>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root StyledTableRowHover-root-91 MuiTableRow-hover"
            style="height: 55px; cursor: initial;"
          >
            <td
              class="MuiTableCell-root StyledTableCell-root-93 MuiTableCell-body MuiTableCell-sizeSmall"
            >
              MortgageApprovalProcess
            </td>
            <td
              class="MuiTableCell-root StyledTableCell-root-93 MuiTableCell-body MuiTableCell-sizeSmall"
            >
              pamAdmin
            </td>
            <td
              class="MuiTableCell-root StyledTableCell-root-93 MuiTableCell-body MuiTableCell-sizeSmall"
            >
              2020-05-12T08:46:15.116
            </td>
            <td
              class="MuiTableCell-root StyledTableCell-root-93 MuiTableCell-body MuiTableCell-sizeSmall"
            >
              <span>
                Qualify
              </span>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root StyledTableRowHover-root-91 MuiTableRow-hover"
            style="height: 55px; cursor: initial;"
          >
            <td
              class="MuiTableCell-root StyledTableCell-root-93 MuiTableCell-body MuiTableCell-sizeSmall"
            >
              MortgageApprovalProcess
            </td>
            <td
              class="MuiTableCell-root StyledTableCell-root-93 MuiTableCell-body MuiTableCell-sizeSmall"
            >
              pamAdmin
            </td>
            <td
              class="MuiTableCell-root StyledTableCell-root-93 MuiTableCell-body MuiTableCell-sizeSmall"
            >
              2020-05-12T10:41:54.418
            </td>
            <td
              class="MuiTableCell-root StyledTableCell-root-93 MuiTableCell-body MuiTableCell-sizeSmall"
            >
              <span>
                Qualify
              </span>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <table
      class="MuiTable-root"
    />
    
  </div>
</div>
`;
